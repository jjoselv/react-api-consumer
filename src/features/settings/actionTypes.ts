export const SET_NATIONALITY = 'SET_NATIONALITY';

export interface SetNationalityAction {
    type: typeof SET_NATIONALITY;
    payload: string;
};
    
export type SETTINGS_ACTION_TYPE = SetNationalityAction;
