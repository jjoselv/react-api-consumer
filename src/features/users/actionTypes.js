export const GET_USERS = 'GET_USERS';
export const GET_NEXT_BATCH = 'GET_NEXT_BATCH';
