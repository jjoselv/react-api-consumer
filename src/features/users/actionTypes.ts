import { User } from "../../components/Users/Users";

export const GET_USERS = 'GET_USERS';
export const GET_NEXT_BATCH = 'GET_NEXT_BATCH';
export type PromiseActions = 'PENDING' | 'FULFILLED' | 'REJECTED';

export interface SetUsersAction {
    type: `${typeof GET_USERS}_${PromiseActions}`;
    payload: {
        data: {
            results: User[];
            info: {
                page: number;
            };
        };
    };
};

export interface SetUsersNextBatchAction {
    type: `${typeof GET_NEXT_BATCH}_${PromiseActions}`;
    payload: {
        data: {
            results: User[];
            info: {
                page: number;
            };
        };
    };
};
    
export type USERS_ACTION_TYPE = SetUsersAction | SetUsersNextBatchAction;
