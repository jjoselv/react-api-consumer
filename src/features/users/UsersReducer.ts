import {User} from '../../components/Users/Users';
import {GET_NEXT_BATCH, GET_USERS, USERS_ACTION_TYPE} from './actionTypes';

export interface UsersState {
  users: undefined | User[],
  isLoading: boolean,
  hasError: boolean,
  isFulfilled: boolean,
  page: number,
}

const initialState: UsersState = {
  users: undefined,
  isLoading: false,
  hasError: false,
  isFulfilled: false,
  page: 1,
};

function reducer(state = initialState, action: USERS_ACTION_TYPE): UsersState {
  switch (action.type) {
    case `${GET_USERS}_PENDING`:
      return {
        ...state,
        isFulfilled: false,
        isLoading: true,
        hasError: false,
        users: undefined,
      };
    case `${GET_USERS}_FULFILLED`:
      return {
        ...state,
        isFulfilled: true,
        isLoading: false,
        hasError: false,
        users: action.payload.data?.results,
        page: action.payload.data?.info?.page,
      };
    case `${GET_USERS}_REJECTED`:
      return {
        ...state,
        isFulfilled: false,
        isLoading: false,
        hasError: true,
        users: undefined,
      };
    case `${GET_NEXT_BATCH}_PENDING`:
      return {
        ...state,
        isFulfilled: false,
        isLoading: true,
        hasError: false,
      };
    case `${GET_NEXT_BATCH}_FULFILLED`:
      return {
        ...state,
        isFulfilled: true,
        isLoading: false,
        hasError: false,
        users: [...(state.users ?? []), ...action.payload.data.results],
        page: action.payload.data?.info?.page,
      };
    case `${GET_NEXT_BATCH}_REJECTED`:
      return {
        ...state,
        isFulfilled: false,
        isLoading: false,
        hasError: true,
      };
    default:
      return state;
  }
};

export default reducer;
