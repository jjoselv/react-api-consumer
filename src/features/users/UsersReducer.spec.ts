import {GET_USERS, GET_NEXT_BATCH} from './actionTypes';
import type {SetUsersAction, SetUsersNextBatchAction, USERS_ACTION_TYPE, PromiseActions} from './actionTypes';

import UsersReducer from './UsersReducer';
import type {User} from '../../components/Users/Users';

describe('features > users > UsersReducer', () => {
  it('returns initial state, if non matched action is dispatched', () => {
    const initialState = {
      users: undefined,
      isLoading: false,
      hasError: false,
      isFulfilled: false,
      page: 0
    };

    const action = {
      type: 'FOO',
    };

    expect(UsersReducer(initialState, action as USERS_ACTION_TYPE)).toBe(initialState);
  });

  it.each([
    [`${GET_USERS}_FULFILLED` as `${typeof GET_USERS}_${PromiseActions}`],
    [`${GET_USERS}_PENDING` as `${typeof GET_USERS}_${PromiseActions}`],
    [`${GET_USERS}_REJECTED` as `${typeof GET_USERS}_${PromiseActions}`],
  ])(`updates state according to dispatched action %s`, actionType => {
    const initialState = {
      users: undefined,
      isLoading: false,
      hasError: false,
      isFulfilled: false,
      page: 1,
    };

    const payload: SetUsersAction["payload"] | undefined =
      actionType === `${GET_USERS}_FULFILLED`
        ? {
            data: {
              results: [{email: 'asdf@asdf.com'} as User],
              info: {
                page: 2,
              },
            },
          }
        : undefined;

    const action = {
      type: actionType,
      payload,
    };

    expect(UsersReducer(initialState, action as SetUsersAction)).toMatchSnapshot();
  });

  it.each([
    [`${GET_NEXT_BATCH}_FULFILLED` as `${typeof GET_NEXT_BATCH}_${PromiseActions}`],
    [`${GET_NEXT_BATCH}_PENDING` as `${typeof GET_NEXT_BATCH}_${PromiseActions}`],
    [`${GET_NEXT_BATCH}_REJECTED` as `${typeof GET_NEXT_BATCH}_${PromiseActions}`],
  ])(`updates state according to dispatched action %s`, actionType => {
    const initialState = {
      users: [{email: "fdsa@ffff.com"}] as User[],
      isLoading: false,
      hasError: false,
      isFulfilled: false,
      page: 1,
    };

    const payload: SetUsersAction["payload"] | undefined =
      actionType === `${GET_NEXT_BATCH}_FULFILLED`
        ? {
            data: {
              results: [{email: 'asdf@asdf.com'} as User],
              info: {
                page: 42,
              },
            },
          }
        : undefined;

    const action = {
      type: actionType,
      payload,
    };

    expect(UsersReducer(initialState, action as SetUsersNextBatchAction)).toMatchSnapshot();
  });
});
