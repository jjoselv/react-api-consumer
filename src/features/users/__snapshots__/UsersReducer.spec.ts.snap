// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features > users > UsersReducer updates state according to dispatched action GET_NEXT_BATCH_FULFILLED 1`] = `
Object {
  "hasError": false,
  "isFulfilled": true,
  "isLoading": false,
  "page": 42,
  "users": Array [
    Object {
      "email": "fdsa@ffff.com",
    },
    Object {
      "email": "asdf@asdf.com",
    },
  ],
}
`;

exports[`features > users > UsersReducer updates state according to dispatched action GET_NEXT_BATCH_PENDING 1`] = `
Object {
  "hasError": false,
  "isFulfilled": false,
  "isLoading": true,
  "page": 1,
  "users": Array [
    Object {
      "email": "fdsa@ffff.com",
    },
  ],
}
`;

exports[`features > users > UsersReducer updates state according to dispatched action GET_NEXT_BATCH_REJECTED 1`] = `
Object {
  "hasError": true,
  "isFulfilled": false,
  "isLoading": false,
  "page": 1,
  "users": Array [
    Object {
      "email": "fdsa@ffff.com",
    },
  ],
}
`;

exports[`features > users > UsersReducer updates state according to dispatched action GET_USERS_FULFILLED 1`] = `
Object {
  "hasError": false,
  "isFulfilled": true,
  "isLoading": false,
  "page": 2,
  "users": Array [
    Object {
      "email": "asdf@asdf.com",
    },
  ],
}
`;

exports[`features > users > UsersReducer updates state according to dispatched action GET_USERS_PENDING 1`] = `
Object {
  "hasError": false,
  "isFulfilled": false,
  "isLoading": true,
  "page": 1,
  "users": undefined,
}
`;

exports[`features > users > UsersReducer updates state according to dispatched action GET_USERS_REJECTED 1`] = `
Object {
  "hasError": true,
  "isFulfilled": false,
  "isLoading": false,
  "page": 1,
  "users": undefined,
}
`;
